day8/alienalphabet.py

```diff
@@ -0,0 +1,131 @@
+##############
+## Solution ##
+##############
+
+def alienAlphabet(orderedWords):
+    # edge cases
+    if len(orderedWords) <= 0:
+        return ''
+    if len(orderedWords) == 1:
+        return removeDups(orderedWords[0])
+
+    # linearization of a DAG
+    graph = createDependencyGraph(orderedWords)
+    alphabet = topSort(graph)
+
+    return "".join(alphabet)
+
+#############
+## Helpers ##
+#############
+
+def createDependencyGraph(orderedWords):
+    graph = {}
+    for i, word in enumerate(orderedWords[:-1]):
+        nextWord = orderedWords[i+1]
+        shorter = min(len(word), len(nextWord))
+        for j in xrange(shorter):
+            letter = word[j]
+            otherLetter = nextWord[j]
+
+            if letter not in graph:
+                graph[letter] = []
+            if otherLetter not in graph:
+                graph[otherLetter] = []
+
+            if letter != otherLetter:
+                graph[letter].append(otherLetter)
+                break
+
+    return graph
+
+# graph is represented as key:value pair
+# key = node
+# value = [list, of, connected, nodes]
+#
+# the graph is also given to be a DAG so all the connected
+# nodes are from outgoing edges
+#
+# this function returns the resulting linearization in a list
+def topSort(graph):
+    sources = getSources(graph)
+    visited = set()
+    postOrderStack = []
+
+    def dfs(start, graph, visitedSet, postOrderStack):
+        visitedSet.add(start)
+
+        # recurse
+        for child in graph[start]:
+            if child not in visitedSet:
+                dfs(child, graph, visitedSet, postOrderStack)
+
+        # done exploring your children, add yourself to the postorder
+        postOrderStack.append(start)        
+
+    # run dfs from all source nodes
+    for node in sources:
```

Comment:
if you actually have multiple sources, this can cause your algorithm to be wrong. take the following dictionary as an example: ['ab', 'ac', 'db', 'df', 'f', 'zc', 'zf']. then your `sources` will be "a" and "b" (you can't prove anything comes before either with this dictionary). then your algorithm will go to source "a" first (same logic applies even if it goes to "b" first) and `postOrderStack` will be ['a', 'd', 'f', 'z']. then this for loop on this line will go to source "b" next and put it on the `postOrderStack` so it'll say ['a', 'd', 'f', 'z', 'b'] which implies 'b' comes after 'f' when actually the opposite is true.

https://en.wikipedia.org/wiki/Topological_sorting has the pseudocode for how to do this DFS